File: tests/benchmark/prime/haskell/prime.hs
-- Compute the Nth prime number

import Data.List

prime :: Int -> Int
prime n = go 0 2 []
  where
    go :: Int -> Int -> [Int] -> Int
    go k p lst =
      if k == n then
        head lst
      else if checkDivisible p lst then
        go k (p + 1) lst
      else
        go (k + 1) (p + 1) (p : lst)

    checkDivisible :: Int -> [Int] -> Bool
    checkDivisible p [] = False
    checkDivisible p (h : _) | p `mod` h == 0 = True
    checkDivisible p (_ : t) = checkDivisible p t

main :: IO ()
main = putStrLn (show (prime (1024 * 16)))

File: tests/benchmark/prime/ocaml/prime.ml
(* Compute the Nth prime *)

let prime n =
  let rec check_divisible p lst =
    match lst with
    | [] -> false
    | h :: t -> if p mod h = 0 then true else check_divisible p t
  in
  let rec go n p lst =
    if n = 0 then
      List.hd lst
    else if check_divisible p lst then
      go n (p + 1) lst
    else
      go (n - 1) (p + 1) (p :: lst)
  in
  go n 2 [];;

print_int (prime (1024 * 16));;
print_newline ();;
